storage/fanout.go

```diff
@@ -240,23 +240,46 @@ func (q *mergeQuerier) Select(params *SelectParams, matchers ...*labels.Matcher)
 func (q *mergeQuerier) SelectSorted(params *SelectParams, matchers ...*labels.Matcher) (SeriesSet, Warnings, error) {
 	seriesSets := make([]SeriesSet, 0, len(q.queriers))
 	var warnings Warnings
+
+	var priErr error = nil
+	type queryResult struct {
+		qr          Querier
+		set         SeriesSet
+		wrn         Warnings
+		selectError error
+	}
+	queryResultChan := make(chan *queryResult)
 	for _, querier := range q.queriers {
-		set, wrn, err := querier.SelectSorted(params, matchers...)
-		q.setQuerierMap[set] = querier
-		if wrn != nil {
-			warnings = append(warnings, wrn...)
+		go func(qr Querier) {
+			if priErr != nil {
+				queryResultChan <- &queryResult{qr: qr, set: nil, wrn: nil, selectError: nil}
+				return
+			}
+			set, wrn, err := qr.SelectSorted(params, matchers...)
+			queryResultChan <- &queryResult{qr: qr, set: set, wrn: wrn, selectError: err}
+		}(querier)
+	}
+	for i := 0; i < len(q.queriers); i++ {
+		qryResult := <-queryResultChan
+		if priErr != nil {
```

Comment:
We may aswell collect all the warnings when there's an error in the primary.

Reply:
done.
![image](https://user-images.githubusercontent.com/9583245/74215125-f712b000-4cda-11ea-9cc6-94069840a22d.png)